src/matching-service/matching.js

```diff
@@ -0,0 +1,181 @@
+
+  await channel.assertQueue(MATCH_REPLY_QUEUE, { durable: false, messageTtl: 30000 });
+
+  clearTimeout(pendingTimeouts[index]);
+
+  const firstUserSocketId = pendingSocketRequests[index];
+  const firstUserId = pendingUserIds[index];
+
+  const secondUserSocketId = socketId;
+  const secondUserId = userId;
+
+  pendingSocketRequests[index] = null;
+  pendingUserIds[index] = null;
+
+  const matchSuccessMessage = {
+    isMatchCancel: false,
+    firstUserSocketId,
+    secondUserSocketId,
+    firstUserId, 
+    secondUserId,
+    complexity: QUESTION_COMPLEXITY_LIST[index]
+  };
+
+  channel.sendToQueue(MATCH_REPLY_QUEUE, Buffer.from(JSON.stringify(matchSuccessMessage)));
+}
+
+
+exports.listenToQueue = async (channel) => {
+  await channel.assertQueue(EASY_QUEUE, { durable: false, messageTtl: 30000 });
+  await channel.assertQueue(MEDIUM_QUEUE, { durable: false, messageTtl: 30000 });
+  await channel.assertQueue(HARD_QUEUE, { durable: false, messageTtl: 30000 });
+
+  const queueNames = [EASY_QUEUE, MEDIUM_QUEUE, HARD_QUEUE];
+  const pendingSocketRequests = [null, null, null];
+  const pendingUserIds= [null, null, null];
+  const pendingTimeouts = [null, null, null];
+
+  for (let index = 0; index < queueNames.length; index++) {
+    const queueName = queueNames[index];
+
+    channel.consume(queueName, (message) => {
+      if (!message) {
+        return;
+      }
+
+      // console.log(message)
+      
+      console.log("consumed queue message");
+
+      const { isCancelRequest, userId, socketId } = JSON.parse(message.content.toString());
+
+      if (isCancelRequest) {
+        handleCancelRequest(channel, index, socketId, userId, pendingSocketRequests, pendingUserIds, pendingTimeouts);
+      } else if (!pendingSocketRequests[index]) {
+        handleSaveMatchRequest(channel, index, socketId, userId, pendingSocketRequests, pendingUserIds, pendingTimeouts);
+      } else {
+        handleSuccessMatchRequest(channel, index, socketId, userId, pendingSocketRequests, pendingUserIds, pendingTimeouts);
+      }
+    }, { noAck: true});
+    
+  }
+}
+
+exports.listenToReplies = async (channel, io) => {
+
+  await channel.assertQueue(MATCH_REPLY_QUEUE, { durable: false, messageTtl: 30000 });
+
+  channel.consume(MATCH_REPLY_QUEUE, async (message) => {
+
+    console.log("received reply")
+
+    const parsedReply = JSON.parse(message.content.toString());
+
+    if (parsedReply.isMatchCancel) {
+      const socketId = parsedReply.firstUserSocketId;
+      io.to(socketId).emit("match-cancelled", "Match cancelled");
+
+      if (io.sockets.sockets.get(socketId)) {
+        io.sockets.sockets.get(socketId).disconnect(true);
+      }
+
+      return;
+    }
+
+    const firstUserSocketId = parsedReply.firstUserSocketId;
+    const firstUserId = parsedReply.firstUserId;
+
+    const secondUserSocketId = parsedReply.secondUserSocketId;
+    const secondUserId = parsedReply.secondUserId;
+
+    const complexity = parsedReply.complexity;
```

Comment:
Can consider destructuring `parsedReply` like we usually do for the other JS objects